### ** "Before analyzing anything, make sure you're solving the RIGHT problem."**

#### 🛍️ Mystery Shopping Analytics
###### Retail Brand Performance Analysis — HS Brands India
---
###### **Analyst:** Thivya T
###### **Tool:** Python Pandas  
###### **Date:** August 2026  
###### **Dataset:** 805 mystery shopping records  
###### **Brands:** Levi's | Tommy Hilfiger | Calvin Klein | Zara | H&M  
###### **Cities:** 10 Indian cities  
---
##### 🎯 Business Goal
###### Identify which brands, cities, and store parameters 
###### require urgent service quality intervention — 
###### and deliver actionable recommendations to the client.
---
##### 📁 Project Structure
- Section 1 — Data Loading & First Look
- Section 2 — Data Quality Audit
- Section 3 — Data Cleaning
- Section 4 — Brand Performance Analysis
- Section 5 — City & Region Analysis
- Section 6 — Store Rankings
- Section 7 — Recommendations

In [1]:
import pandas as pd
import numpy as np

#### Section 1 — Data Loading & First Look
---
**Purpose:** Understand what data we have before any analysis.  
**Method:** Load dataset → Check shape → Check columns → Check data types → Check missing values

In [2]:
Mystery_shop_data = pd.read_csv("mystery_shopping_raw (1).csv")

In [3]:
Mystery_shop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 805 entries, 0 to 804
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   audit_id                   805 non-null    object 
 1   store_id                   805 non-null    object 
 2   brand                      805 non-null    object 
 3   city                       805 non-null    object 
 4   region                     805 non-null    object 
 5   audit_date                 805 non-null    object 
 6   shopper_name               805 non-null    object 
 7   visit_type                 805 non-null    object 
 8   manager_present            805 non-null    object 
 9   greeting_score             768 non-null    object 
 10  product_knowledge_score    768 non-null    object 
 11  store_cleanliness_score    769 non-null    object 
 12  staff_appearance_score     767 non-null    object 
 13  billing_efficiency_score   766 non-null    object 

In [4]:
print("Shape:", Mystery_shop_data.shape)

Shape: (805, 22)


In [5]:
Mystery_shop_data.isnull().sum()

audit_id                       0
store_id                       0
brand                          0
city                           0
region                         0
audit_date                     0
shopper_name                   0
visit_type                     0
manager_present                0
greeting_score                37
product_knowledge_score       37
store_cleanliness_score       36
staff_appearance_score        38
billing_efficiency_score      39
customer_engagement_score     42
complaint_handling_score      26
upselling_score               28
queue_management_score        41
overall_experience_score      46
audit_fee                     45
visit_difficulty             220
shopper_comments             290
dtype: int64

In [6]:
Mystery_shop_data.head(6)

,audit_id,store_id,brand,city,region,audit_date,shopper_name,visit_type,manager_present,greeting_score,...,staff_appearance_score,billing_efficiency_score,customer_engagement_score,complaint_handling_score,upselling_score,queue_management_score,overall_experience_score,audit_fee,visit_difficulty,shopper_comments
0,AUD0001,STR013,ZARA,Kolkata,East,12-09-2023,Ananya Singh,ANNOUNCED,No,6.3,...,8.5,6.3,6.4,7.6,7.7,6.3,2.9,1565,2.0,Great experience
1,AUD0002,STR006,H&m,Pune,West,03/31/2023,PRIYA kumar,announced,0,5.9,...,6.2,6.9,0,6.6,4,7.5,NaN,1129,5.0,Store was clean
2,AUD0003,STR008,Tommy Hilfiger,Mumbai,West,12-01-2024,Ananya Mehta,ANNOUNCED,Yes,0,...,6,NaN,7.7,5.7,7.2,5.5,NaN,1917,3.0,NaN
3,AUD0004,STR048,levis,Bangalore,South,06-04-2023,ANANYA PATEL,Unannounced,FALSE,6.9,...,6,7.1,7.4,8.6,6.3,7.3,10,1540,5.0,Excellent service
4,AUD0005,STR025,Calvin klein,Chennai,South,02/13/2024,Karthik SINGH,Announced,NO,8.1,...,8.2,4.9,7.1,4.3,8.8,10,6.7,1436,2.0,NaN
5,AUD0006,STR019,Tommy Hilfiger,Kolkata,East,03/24/2024,ANANYA Patel,unannounced,1,9.4,...,7.6,5.6,6.1,8.4,5.3,6.4,6.6,1126,NaN,NaN


#### Inconsistent brand names

##### If inconsistent brand names are not fixed, the same brand may be treated as multiple different brands during analysis, resulting in incorrect calculations and misleading brand-level insights. It can also affect the accuracy and professionalism of client-facing reports, as incorrect spelling or formatting of a client's brand name may be unacceptable.

In [89]:
Mystery_shop_data["brand"].unique()

array(['ZARA ', 'H&m', 'Tommy Hilfiger', 'levis', 'Calvin klein',
       'Tommy Hilfiger ', 'h&m', 'TOMMY HILFIGER', 'Zara ', "Levi's",
       'Levis', 'H&M', 'ZARA', 'Calvin  Klein', 'tommy hilfiger',
       "LEVI'S", 'HM', 'Calvin Klein', 'CALVIN KLEIN', 'CK', 'Levi s',
       'Tommy', 'zara', 'H & M', 'Zara'], dtype=object)

In [7]:
Brand_map={'ZARA ':'Zara','Zara ':'Zara','zara':'Zara','ZARA':'Zara',
           'h&m':'H&M','H&m':'H&M','H & M':'H&M','HM':'H&M',
           'Tommy':'Tommy Hilfiger','tommy hilfiger':'Tommy Hilfiger','TOMMY HILFIGER':'Tommy Hilfiger','Tommy Hilfiger ':'Tommy Hilfiger',
          'levis':"Levi's",'Levis':"Levi's","LEVI'S":"Levi's",'Levi s':"Levi's",
          'CALVIN KLEIN':'Calvin Klein','CK':'Calvin Klein','Calvin  Klein':'Calvin Klein','Calvin klein':'Calvin Klein'}

In [8]:
Mystery_shop_data['brand'] = Mystery_shop_data['brand'].replace(Brand_map)

In [9]:
Mystery_shop_data["shopper_name"] = Mystery_shop_data["shopper_name"].str.title().str.strip()
Mystery_shop_data["visit_type"] = Mystery_shop_data["visit_type"].str.title().str.strip()
Mystery_shop_data["manager_present"] = Mystery_shop_data["manager_present"].str.title().str.strip()

In [10]:
Mystery_shop_data["manager_present"] = Mystery_shop_data["manager_present"].replace("False", "No")
Mystery_shop_data["manager_present"] = Mystery_shop_data["manager_present"].replace("True", "Yes")
Mystery_shop_data["manager_present"] = Mystery_shop_data["manager_present"].replace("1", "Yes")
Mystery_shop_data["manager_present"] = Mystery_shop_data["manager_present"].replace("0", "No")

In [164]:
Mystery_shop_data["visit_type"] = Mystery_shop_data["visit_type"].replace("Un-announced", "Unannounced")
Mystery_shop_data["visit_type"] = Mystery_shop_data["visit_type"].replace("Un-Announced", "Unannounced")
Mystery_shop_data["visit_type"].value_counts()

visit_type
Unannounced    405
Announced      395
Name: count, dtype: int64

In [11]:
Mystery_shop_data["manager_present"].value_counts()

manager_present
Yes    403
No     402
Name: count, dtype: int64

In [12]:
Mystery_shop_data["shopper_name"].value_counts()

shopper_name
Vikram Kumar     23
Priya Kumar      21
Priya Singh      21
Vikram Sharma    20
Rahul Patel      18
                 ..
Meena Nair        2
Meena Iyer        2
Arjun Gupta       1
Arjun Sharma      1
Meena Gupta       1
Name: count, Length: 100, dtype: int64

In [13]:
Mystery_shop_data["brand"].unique()

array(['Zara', 'H&M', 'Tommy Hilfiger', "Levi's", 'Calvin Klein'],
      dtype=object)

In [14]:
Mystery_shop_data["audit_id"].isnull().sum()

np.int64(0)

In [15]:
Mystery_shop_data[Mystery_shop_data.duplicated()]

,audit_id,store_id,brand,city,region,audit_date,shopper_name,visit_type,manager_present,greeting_score,...,staff_appearance_score,billing_efficiency_score,customer_engagement_score,complaint_handling_score,upselling_score,queue_management_score,overall_experience_score,audit_fee,visit_difficulty,shopper_comments
800,AUD0042,STR001,H&M,Bangalore,South,11/21/2023,Priya Singh,Unannounced,Yes,8.1,...,8,4.7,6.1,5.8,7.1,8.4,5.6,1714,5.0,Staff was helpful
801,AUD0255,STR007,Zara,Jaipur,North,18-03-2023,Priya Kumar,Un-Announced,Yes,5.8,...,8.1,8.4,7.8,4.6,5.8,5.8,7,NaN,NaN,Long queue observed
802,AUD0293,STR027,Calvin Klein,Hyderabad,South,19-01-2023,Rahul Patel,Unannounced,No,-,...,8.9,NaN,0,9,6.2,5.7,7.2,1453,5.0,Billing took too long
803,AUD0222,STR036,Levi's,Ahmedabad,West,09-12-2024,Meena Reddy,Un-Announced,No,6.6,...,5.3,6.4,0,9.9,5.7,5.7,7.5,1531,3.0,Staff was rude
804,AUD0279,STR015,Tommy Hilfiger,Coimbatore,South,19-09-2024,Rahul Reddy,Unannounced,Yes,6.5,...,7,0,7.3,5.8,9.2,6.7,8.9,837,NaN,NaN


In [16]:
Mystery_shop_data.drop_duplicates(inplace=True)

In [17]:
Mystery_shop_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 800 entries, 0 to 799
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   audit_id                   800 non-null    object 
 1   store_id                   800 non-null    object 
 2   brand                      800 non-null    object 
 3   city                       800 non-null    object 
 4   region                     800 non-null    object 
 5   audit_date                 800 non-null    object 
 6   shopper_name               800 non-null    object 
 7   visit_type                 800 non-null    object 
 8   manager_present            800 non-null    object 
 9   greeting_score             763 non-null    object 
 10  product_knowledge_score    763 non-null    object 
 11  store_cleanliness_score    764 non-null    object 
 12  staff_appearance_score     762 non-null    object 
 13  billing_efficiency_score   762 non-null    object 
 14 

In [30]:
pd.to_numeric(Mystery_shop_data["greeting_score"] ,errors='coerce')

0      6.300000
1      5.900000
2      6.792481
3      6.900000
4      8.100000
         ...   
795    9.300000
796    6.467187
797    6.600000
798    6.792481
799    5.700000
Name: greeting_score, Length: 800, dtype: float64

In [25]:
score_cols = [
    'greeting_score',
    'product_knowledge_score',
    'store_cleanliness_score',
    'staff_appearance_score',
    'billing_efficiency_score',
    'customer_engagement_score',
    'complaint_handling_score',
    'upselling_score',
    'queue_management_score',
    'overall_experience_score'
]

In [26]:
for col in score_cols:
    Mystery_shop_data[col] = pd.to_numeric(
        Mystery_shop_data[col],
        errors='coerce'
    )

In [27]:
Mystery_shop_data[score_cols].dtypes

greeting_score               float64
product_knowledge_score      float64
store_cleanliness_score      float64
staff_appearance_score       float64
billing_efficiency_score     float64
customer_engagement_score    float64
complaint_handling_score     float64
upselling_score              float64
queue_management_score       float64
overall_experience_score     float64
dtype: object

In [28]:
for col in score_cols:
    Mystery_shop_data[col] = Mystery_shop_data[col].where( Mystery_shop_data[col].between(1, 10),np.nan) 
    
    ### series.where(condition, replacement)

In [29]:
for col in score_cols:
    Mystery_shop_data[col] = Mystery_shop_data[col].fillna(Mystery_shop_data.groupby('brand')[col].transform('mean'))

In [110]:
Mystery_shop_data[score_cols].isnull().sum()

greeting_score               0
product_knowledge_score      0
store_cleanliness_score      0
staff_appearance_score       0
billing_efficiency_score     0
customer_engagement_score    0
complaint_handling_score     0
upselling_score              0
queue_management_score       0
overall_experience_score     0
dtype: int64

In [113]:
Mystery_shop_data["audit_fee"]

0      1565
1      1129
2      1917
3      1540
4      1436
       ... 
795    1610
796    1241
797    1753
798    1060
799    1882
Name: audit_fee, Length: 800, dtype: object

In [112]:
Mystery_shop_data['audit_fee'] = Mystery_shop_data['audit_fee'].str.replace('Rs.', '')

In [118]:
### Why errors='coerce'?If any unexpected non-numeric value still exists, such as:'N/A',NA','-',normal pd.to_numeric() will give an error. 
### errors='coerce' invalid text is converted to NaN.

Mystery_shop_data['audit_fee'] = pd.to_numeric(Mystery_shop_data['audit_fee'],errors='coerce')

In [119]:
Mystery_shop_data["audit_fee"].dtype

dtype('float64')

In [120]:
Mystery_shop_data['audit_fee'].isnull().sum()

np.int64(44)

In [ ]:
# ── Pre-cleaning Investigation ──────────────────
# How many negative fees exist?
negative_fees = Mystery_shop_data[
    Mystery_shop_data['audit_fee'] < 0]

print(f"Negative audit fees found: {len(negative_fees)}")
print(f"Affected records:")
print(negative_fees[['audit_id','store_id',
                      'brand','audit_fee']])

In [123]:
### 3. Replace negative values with NaN

Mystery_shop_data['audit_fee'] = Mystery_shop_data['audit_fee'].where(Mystery_shop_data['audit_fee'] >= 0,np.nan)

In [126]:
Mystery_shop_data.head(5)

,audit_id,store_id,brand,city,region,audit_date,shopper_name,visit_type,manager_present,greeting_score,...,staff_appearance_score,billing_efficiency_score,customer_engagement_score,complaint_handling_score,upselling_score,queue_management_score,overall_experience_score,audit_fee,visit_difficulty,shopper_comments
0,AUD0001,STR013,Zara,Kolkata,East,12-09-2023,Ananya Singh,Announced,No,6.300000,...,8.5,6.300000,6.400000,7.6,7.7,6.3,2.900000,1565.0,2.0,Great experience
1,AUD0002,STR006,H&M,Pune,West,03/31/2023,Priya Kumar,Announced,No,5.900000,...,6.2,6.900000,6.153797,6.6,4.0,7.5,6.309434,1129.0,5.0,Store was clean
2,AUD0003,STR008,Tommy Hilfiger,Mumbai,West,12-01-2024,Ananya Mehta,Announced,Yes,6.792481,...,6.0,6.908462,7.700000,5.7,7.2,5.5,7.175781,1917.0,3.0,NaN
3,AUD0004,STR048,Levi's,Bangalore,South,06-04-2023,Ananya Patel,Unannounced,No,6.900000,...,6.0,7.100000,7.400000,8.6,6.3,7.3,10.000000,1540.0,5.0,Excellent service
4,AUD0005,STR025,Calvin Klein,Chennai,South,02/13/2024,Karthik Singh,Announced,No,8.100000,...,8.2,4.900000,7.100000,4.3,8.8,10.0,6.700000,1436.0,2.0,NaN


In [125]:
Mystery_shop_data['audit_fee'] = Mystery_shop_data['audit_fee'].fillna(
    Mystery_shop_data['audit_fee'].median()
)

In [130]:
##Before: "NA"     ← text pretending to represent missing data
## After:  NaN      ← Pandas officially recognizes it as missing
Mystery_shop_data['shopper_comments'] = Mystery_shop_data['shopper_comments'].replace(['', ' ', 'NA', 'N/A'], np.nan)

In [132]:
Mystery_shop_data['shopper_comments'].isna().sum()

np.int64(345)

In [41]:
Mystery_shop_data['audit_date'].dt.year

0      2023
1      2023
2      2024
3      2023
4      2024
       ... 
795    2023
796    2024
797    2023
798    2024
799    2023
Name: audit_date, Length: 800, dtype: int32

In [42]:
Mystery_shop_data['audit_date'].dt.month

0      12
1       3
2      12
3       6
4       2
       ..
795     3
796     1
797     1
798     9
799     9
Name: audit_date, Length: 800, dtype: int32

In [43]:
Mystery_shop_data['audit_date'].dt.day

0       9
1      31
2       1
3       4
4      13
       ..
795     8
796     9
797     2
798    21
799    24
Name: audit_date, Length: 800, dtype: int32

In [36]:
Mystery_shop_data['audit_date'] = pd.to_datetime(
    Mystery_shop_data['audit_date'],
    format='mixed'
)

In [37]:
Mystery_shop_data['audit_date'].dtype

dtype('<M8[ns]')

In [38]:
Mystery_shop_data['year'] = Mystery_shop_data['audit_date'].dt.year

Mystery_shop_data['month'] = Mystery_shop_data['audit_date'].dt.month

In [39]:
Mystery_shop_data[['audit_date', 'year', 'month']].head()

,audit_date,year,month
0,2023-12-09,2023,12
1,2023-03-31,2023,3
2,2024-12-01,2024,12
3,2023-06-04,2023,6
4,2024-02-13,2024,2


In [143]:
Mystery_shop_data.head(10)

,audit_id,store_id,brand,city,region,audit_date,shopper_name,visit_type,manager_present,greeting_score,...,customer_engagement_score,complaint_handling_score,upselling_score,queue_management_score,overall_experience_score,audit_fee,visit_difficulty,shopper_comments,year,month
0,AUD0001,STR013,Zara,Kolkata,East,2023-12-09,Ananya Singh,Announced,No,6.30,...,6.40,7.6,7.70,6.3,2.90,1565.0,2.0,Great experience,2023,12
1,AUD0002,STR006,H&M,Pune,West,2023-03-31,Priya Kumar,Announced,No,5.90,...,6.15,6.6,4.00,7.5,6.31,1129.0,5.0,Store was clean,2023,3
2,AUD0003,STR008,Tommy Hilfiger,Mumbai,West,2024-12-01,Ananya Mehta,Announced,Yes,6.79,...,7.70,5.7,7.20,5.5,7.18,1917.0,3.0,NaN,2024,12
3,AUD0004,STR048,Levi's,Bangalore,South,2023-06-04,Ananya Patel,Unannounced,No,6.90,...,7.40,8.6,6.30,7.3,10.00,1540.0,5.0,Excellent service,2023,6
4,AUD0005,STR025,Calvin Klein,Chennai,South,2024-02-13,Karthik Singh,Announced,No,8.10,...,7.10,4.3,8.80,10.0,6.70,1436.0,2.0,NaN,2024,2
5,AUD0006,STR019,Tommy Hilfiger,Kolkata,East,2024-03-24,Ananya Patel,Unannounced,Yes,9.40,...,6.10,8.4,5.30,6.4,6.60,1126.0,NaN,NaN,2024,3
6,AUD0007,STR013,H&M,Ahmedabad,West,2023-07-22,Karthik Reddy,Unannounced,Yes,6.27,...,7.70,10.0,7.40,7.0,5.60,1636.0,1.0,Long queue observed,2023,7
7,AUD0008,STR034,Tommy Hilfiger,Mumbai,West,2023-06-21,Priya Patel,Unannounced,No,6.40,...,5.20,8.8,6.30,5.9,6.10,1015.0,5.0,Staff was helpful,2023,6
8,AUD0009,STR024,Zara,Hyderabad,South,2024-08-25,Priya Patel,Announced,Yes,5.60,...,7.60,7.5,4.00,6.8,4.80,1596.0,3.0,Excellent service,2024,8
9,AUD0010,STR045,Zara,Coimbatore,South,2024-05-05,Ananya Singh,Unannounced,Yes,6.10,...,5.60,6.2,6.38,7.9,7.80,1993.0,3.0,Store was clean,2024,5


In [32]:
Mystery_shop_data[score_cols] = Mystery_shop_data[score_cols].round(2)

In [165]:
print("=" * 45)
print("POST-CLEANING VERIFICATION REPORT")
print("=" * 45)

print(f"\n✅ Total Records: {len(Mystery_shop_data)}")

print(f"\n✅ Unique Brands: {Mystery_shop_data['brand'].nunique()}")
print(Mystery_shop_data['brand'].value_counts())

print(f"\n✅ Unique Visit Types:")
print(Mystery_shop_data['visit_type'].value_counts())

print(f"\n✅ Manager Present Values:")
print(Mystery_shop_data['manager_present'].value_counts())

print(f"\n✅ Remaining Missing Values:")
print(Mystery_shop_data.isnull().sum()[
      Mystery_shop_data.isnull().sum() > 0])

print(f"\n✅ Score Ranges After Cleaning:")
print(Mystery_shop_data[score_cols].agg(['min','max']).round(2))

print("=" * 45)
print("DATA IS CLEAN — READY FOR ANALYSIS")
print("=" * 45)

POST-CLEANING VERIFICATION REPORT

✅ Total Records: 800

✅ Unique Brands: 5
brand
H&M               179
Levi's            170
Calvin Klein      161
Tommy Hilfiger    146
Zara              144
Name: count, dtype: int64

✅ Unique Visit Types:
visit_type
Unannounced    405
Announced      395
Name: count, dtype: int64

✅ Manager Present Values:
manager_present
No     400
Yes    400
Name: count, dtype: int64

✅ Remaining Missing Values:
audit_fee            44
visit_difficulty    218
shopper_comments    289
dtype: int64

✅ Score Ranges After Cleaning:
     greeting_score  product_knowledge_score  store_cleanliness_score  \
min             2.6                      1.0                      1.8   
max            10.0                     10.0                     10.0   

     staff_appearance_score  billing_efficiency_score  \
min                     1.4                       1.6   
max                    10.0                      10.0   

     customer_engagement_score  complaint_handling_scor

In [158]:
print(f"\n✅ Unique Visit Types:")
print(Mystery_shop_data['visit_type'].value_counts())


✅ Unique Visit Types:
visit_type
Announced       395
Unannounced     276
Un-Announced    129
Name: count, dtype: int64


##### 🎤 Interview Q: 'How do you handle messy data?' Answer: Walk through your 10-step cleaning process above. 
Before any analysis — I cleaned 800+ mystery shopping records resolving 12 data quality issues: inconsistent brand names, mixed date formats, text in numeric columns, out-of-range scores, negative fees, and duplicate records. Every cleaning decision was documented with business justification — because clean data is the foundation of trustworthy analysis.

In [34]:
Mystery_shop_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 800 entries, 0 to 799
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   audit_id                   800 non-null    object 
 1   store_id                   800 non-null    object 
 2   brand                      800 non-null    object 
 3   city                       800 non-null    object 
 4   region                     800 non-null    object 
 5   audit_date                 800 non-null    object 
 6   shopper_name               800 non-null    object 
 7   visit_type                 800 non-null    object 
 8   manager_present            800 non-null    object 
 9   greeting_score             800 non-null    float64
 10  product_knowledge_score    800 non-null    float64
 11  store_cleanliness_score    800 non-null    float64
 12  staff_appearance_score     800 non-null    float64
 13  billing_efficiency_score   800 non-null    float64
 14 

##### Question 1 — Brand Performance Overview:
• Calculate average score per brand across ALL 10 parameters
• Create an overall_score = mean of all 10 parameter scores
• Rank brands from highest to lowest overall score
• Business question: Which brand delivers the best customer experience?
 

In [51]:
score_cols = [
    'greeting_score',
    'product_knowledge_score',
    'store_cleanliness_score',
    'staff_appearance_score',
    'billing_efficiency_score',
    'customer_engagement_score',
    'complaint_handling_score',
    'upselling_score',
    'queue_management_score',
    'overall_experience_score'
]


In [67]:
# Step 1: Create overall score for EACH AUDIT ROW

Mystery_shop_data["overall_score"] = (
    Mystery_shop_data[score_cols]
    .mean(axis=1).round(2)
)
Mystery_shop_data["overall_score"]

0      6.85
1      6.28
2      6.55
3      7.27
4      7.40
       ... 
795    6.97
796    6.43
797    7.47
798    6.72
799    7.25
Name: overall_score, Length: 800, dtype: float64

In [68]:
# Step 2: Calculate average overall score for EACH BRAND
Overall_average = (
    Mystery_shop_data
    .groupby("brand")["overall_score"]
    .mean()
    .reset_index(name="Overall_average")
    .round(2)
)

Overall_average

,brand,Overall_average
0,Calvin Klein,7.52
1,H&M,6.23
2,Levi's,7.24
3,Tommy Hilfiger,6.85
4,Zara,6.50


In [72]:
Overall_average = Overall_average.sort_values(
    by="Overall_average",
    ascending=False)

Overall_average

,brand,Overall_average
0,Calvin Klein,7.52
1,Levi's,7.24
2,Tommy Hilfiger,6.85
3,Zara,6.50
4,H&M,6.23


Which brand delivers the best customer experience? : Calvin Klein with an Overall CX average of 7.52

#### Question 2 — Parameter Analysis:
• Calculate average score per parameter across all brands
• Identify the 3 weakest parameters company-wide
• Business question: Where should training focus?


In [85]:
Mystery_shop_data.groupby("brand")[score_cols].mean().round(2)

,greeting_score,product_knowledge_score,store_cleanliness_score,staff_appearance_score,billing_efficiency_score,customer_engagement_score,complaint_handling_score,upselling_score,queue_management_score,overall_experience_score
brand,,,,,,,,,,
Calvin Klein,7.45,7.68,7.70,7.66,7.46,7.40,7.36,7.54,7.69,7.26
H&M,6.27,6.03,6.20,6.16,6.06,6.15,6.43,6.36,6.33,6.31
Levi's,7.30,7.13,7.10,7.40,7.34,7.40,7.20,7.28,7.19,7.07
Tommy Hilfiger,6.79,6.72,6.61,7.07,6.91,6.95,6.73,6.83,6.69,7.18
Zara,6.47,6.60,6.64,6.60,6.59,6.52,6.33,6.38,6.49,6.43


In [99]:
Parameter_average = (
    Mystery_shop_data[score_cols]
    .mean()
    .round(2)    
)
Parameter_average

greeting_score               6.86
product_knowledge_score      6.82
store_cleanliness_score      6.85
staff_appearance_score       6.97
billing_efficiency_score     6.86
customer_engagement_score    6.88
complaint_handling_score     6.82
upselling_score              6.88
queue_management_score       6.88
overall_experience_score     6.84
dtype: float64

In [96]:
Parameter_average = (
    Mystery_shop_data[score_cols]
    .mean()
    .round(2)
    .reset_index()
   
)
Parameter_average

,index,0
0,greeting_score,6.86
1,product_knowledge_score,6.82
2,store_cleanliness_score,6.85
3,staff_appearance_score,6.97
4,billing_efficiency_score,6.86
5,customer_engagement_score,6.88
6,complaint_handling_score,6.82
7,upselling_score,6.88
8,queue_management_score,6.88
9,overall_experience_score,6.84


In [101]:
Parameter_average = (
    Mystery_shop_data[score_cols]
    .mean()
    .round(2)
    .reset_index()
    .rename(columns={
        "index": "Parameter",
        0: "Parameter_average"
    })
)
Parameter_average

,Parameter,Parameter_average
0,greeting_score,6.86
1,product_knowledge_score,6.82
2,store_cleanliness_score,6.85
3,staff_appearance_score,6.97
4,billing_efficiency_score,6.86
5,customer_engagement_score,6.88
6,complaint_handling_score,6.82
7,upselling_score,6.88
8,queue_management_score,6.88
9,overall_experience_score,6.84


In [102]:
weakest_3 = (Parameter_average.sort_values(by="Parameter_average",ascending=True).head(3))
weakest_3

,Parameter,Parameter_average
1,product_knowledge_score,6.82
6,complaint_handling_score,6.82
9,overall_experience_score,6.84


In [107]:
Mystery_shop_data[score_cols].mean().nsmallest(3)

complaint_handling_score    6.818563
product_knowledge_score     6.821962
overall_experience_score    6.842338
dtype: float64

#### Question 3 — City & Region Performance:
• Calculate average overall score by city
• Calculate average overall score by region
• Identify cities needing urgent intervention (bottom 3)
• Business question: Which geography needs most attention?

In [133]:
Average_city_CX_score = (Mystery_shop_data.groupby("city")[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "city_average"}))
Average_city_CX_score

,city,city_average
0,Ahmedabad,7.02
1,Bangalore,6.88
2,Chennai,6.97
3,Coimbatore,6.96
4,Delhi,6.75
5,Hyderabad,6.77
6,Jaipur,6.86
7,Kolkata,6.79
8,Mumbai,6.88
9,Pune,6.79


In [142]:
Bottom_cities =Average_city_CX_score.sort_values("city_average",ascending=True).head(3)
Bottom_cities

,city,city_average
4,Delhi,6.75
5,Hyderabad,6.77
7,Kolkata,6.79


In [131]:
Average_Regional_CX_score = (Mystery_shop_data.groupby("region")[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "Regional_average"}))
Average_Regional_CX_score

,region,Regional_average
0,East,6.79
1,North,6.81
2,South,6.89
3,West,6.90


In [141]:
Bottom_regions= Average_Regional_CX_score.sort_values("Regional_average",ascending=True).head(4)
Bottom_regions

,region,Regional_average
0,East,6.79
1,North,6.81
2,South,6.89
3,West,6.90


#### Which geography needs most attention? - Delhi,Hyderabad & Kolkata

#### Question 4 — Store Rankings:
• Calculate average overall score per store_id
• Identify Top 5 and Bottom 5 stores
• Which brand do bottom 5 stores belong to?
• Business question: Which stores need immediate action?


In [150]:
Average_store_CX_score = (Mystery_shop_data.groupby(["store_id","brand"])[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "Store_average"}).head(10))
Average_store_CX_score

,store_id,brand,Store_average
0,STR001,Calvin Klein,7.64
1,STR001,H&M,6.19
2,STR001,Levi's,7.25
3,STR001,Tommy Hilfiger,5.85
4,STR001,Zara,6.44
5,STR002,Calvin Klein,8.19
6,STR002,H&M,5.94
7,STR002,Levi's,7.21
8,STR002,Tommy Hilfiger,6.60
9,STR002,Zara,6.09


In [151]:
Bottom_stores= Average_store_CX_score.sort_values("Store_average",ascending=True).head(5)
Bottom_stores

,store_id,brand,Store_average
3,STR001,Tommy Hilfiger,5.85
6,STR002,H&M,5.94
9,STR002,Zara,6.09
1,STR001,H&M,6.19
4,STR001,Zara,6.44


In [152]:
Top_stores= Average_store_CX_score.sort_values("Store_average",ascending=False).head(5)
Top_stores

,store_id,brand,Store_average
5,STR002,Calvin Klein,8.19
0,STR001,Calvin Klein,7.64
2,STR001,Levi's,7.25
7,STR002,Levi's,7.21
8,STR002,Tommy Hilfiger,6.60


#### • Business question: Which stores need immediate action? HM and Zara

#### Question 5 — Visit Type Impact:
• Compare scores for Announced vs Unannounced visits
• Business question: Do stores perform better when they know the auditor is coming?
• Business insight: A big gap = staff are not maintaining standards consistently


In [166]:
Mystery_shop_data.groupby("visit_type")[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "visit_type_average"})

,visit_type,visit_type_average
0,Announced,6.84
1,Unannounced,6.90


### Business question: Do stores perform better when they know the auditor is coming? slightly better when announced

#### Question 6 — Manager Presence Impact:
• Compare scores when manager was present vs absent
• Business question: Does manager presence improve service quality?

In [167]:
Mystery_shop_data.groupby("manager_present")[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "MG_Presence_type_average"})

,manager_present,MG_Presence_type_average
0,No,6.88
1,Yes,6.85


##### Compare scores when manager was present vs absent -intrestingly in the abscence of manager the scores is higher slightly/ No Manager Presence doesn't affect service quality

#### Question 7 — Trend Analysis:
• Calculate monthly average overall score
• Compare 2023 vs 2024 performance
• Business question: Is performance improving or declining over time?


In [172]:
Mystery_shop_data.groupby("month")[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "Monthly_average"}).sort_values("Monthly_average")

,month,Monthly_average
10,11,6.77
0,1,6.80
2,3,6.82
3,4,6.84
9,10,6.84
8,9,6.87
11,12,6.87
7,8,6.87
5,6,6.88
1,2,6.91


In [171]:
Mystery_shop_data.groupby("year")[score_cols].mean().mean(axis=1).round(2).reset_index().rename(columns={0: "Yearly_average"})

,year,Yearly_average
0,2023,6.87
1,2024,6.86


### Is performance improving or declining over time? No , slightly declining

##### ## Section 7 — Key Findings & Recommendations
---

### 🏆 Key Findings

**Finding 1 — Brand Performance:**
Calvin Klein leads with highest overall CX score of 8.19
H&M and Zara are consistent underperformers across 
multiple parameters and cities.

**Finding 2 — Weakest Parameters Company-Wide:**
Three parameters consistently score below average:
- Complaint Handling: 6.82 — staff lack conflict resolution skills
- Product Knowledge: 6.82 — staff cannot answer customer queries
- Overall Experience: 6.84 — reflects cumulative service gaps

**Finding 3 — Geography Risk:**
Delhi, Hyderabad, and Kolkata require urgent intervention.
These cities show below-average scores across multiple brands
suggesting regional training gaps — not brand-specific issues.

**Finding 4 — Store-Level Risk:**
STR001 and STR002 are bottom performing stores.
Both require immediate store-level audit and action plan.

**Finding 5 — Announced vs Unannounced Gap:**
Scores are slightly higher during announced visits.
This indicates staff are not maintaining consistent 
service standards — they perform better only when 
they know they are being evaluated.

**Finding 6 — Manager Presence:**
Interestingly, scores are marginally higher when 
manager is absent. Staff appear to work more naturally 
and confidently without supervisory pressure.
Manager presence alone does not guarantee 
better service quality.

**Finding 7 — Performance Trend:**
Overall performance has marginally declined from 
6.87 (2023) to 6.86 (2024). While the decline is small,
the direction is concerning — intervention is needed
before it becomes a significant drop.

---

### 🎯 Recommendations

**Recommendation 1 — IMMEDIATE (Within 1 Month):**
Deploy targeted store audits at STR001 and STR002.
These bottom-performing stores need a physical 
inspection, staff interview, and a 30-day 
improvement plan with weekly check-ins.

**Recommendation 2 — SHORT TERM (Within 3 Months):**
Launch complaint handling and product knowledge 
training programs across all H&M and Zara stores 
in Delhi, Hyderabad, and Kolkata — the three cities 
with lowest performance scores.
These two skills account for the widest scoring gaps.

**Recommendation 3 — MEDIUM TERM (Within 6 Months):**
Increase frequency of unannounced mystery shopping 
visits from current ratio to 70% unannounced, 30% 
announced. This will give a more accurate picture 
of true daily service standards and reduce the 
performance gap between announced and unannounced visits.

**Recommendation 4 — BRAND SPECIFIC:**
H&M and Zara brand managers should review their 
staff training curriculum. Calvin Klein's training 
approach — which produces the highest CX scores — 
should be benchmarked and adapted for H&M and Zara.

**Recommendation 5 — MANAGEMENT APPROACH:**
Rather than increasing supervisory presence on 
shop floors, invest in building staff confidence 
and autonomy. Data shows staff perform slightly 
better without direct supervision — suggesting 
empowerment-based management works better than 
control-based supervision in retail environments.

**Recommendation 6 — TREND ALERT:**
The year-on-year decline — though marginal — 
must be arrested immediately. A quarterly 
performance review meeting between HS Brands 
and each brand manager is recommended to track 
progress against these recommendations.

---

### 💼 Overall Conclusion

> "Calvin Klein leads retail CX performance in India. 
> H&M and Zara require urgent brand-level intervention, 
> particularly in Delhi, Hyderabad, and Kolkata. 
> Staff training in complaint handling and product 
> knowledge, combined with increased unannounced 
> auditing, will be the highest-impact interventions 
> to reverse the declining performance trend."